code/game/machinery/dance_machine.dm

```diff
@@ -0,0 +1,514 @@
+				lights_spin()
+				updateUsrDialog()
+			else if(active)
+				active = FALSE
+				STOP_PROCESSING(SSobj, src)
+				update_icon()
+				dance_over()
+				stop = world.time + 300
+				updateUsrDialog()
+		if("select")
+			if(active)
+				to_chat(usr, "<span class='warning'>Error: You cannot change the song until the current one is over.</span>")
+				return
+			check_GBP()
+			select_name = input(usr, "Choose your song", "Track:") as null|anything in available
+			if (QDELETED(src))
+				return
+			for(var/datum/track/S in songs)
+				if(select_name == S.song_name)
+					selection = S
+					break
+			updateUsrDialog()
+		if("horn")
+			deejay('sound/items/AirHorn2.ogg')
+		if("alert")
+			deejay('sound/misc/notice1.ogg')
+		if("siren")
+			deejay('sound/machines/engine_alert1.ogg')
+		if("honk")
+			deejay('sound/items/bikehorn.ogg')
+		if("pump")
+			deejay('sound/weapons/shotgunpump.ogg')
+		if("pop")
+			deejay('sound/weapons/Gunshot3.ogg')
+		if("saber")
+			deejay('sound/weapons/saberon.ogg')
+		if("harm")
+			deejay('sound/AI/harmalarm.ogg')
+
+/obj/machinery/disco/proc/deejay(var/S)
+	if (QDELETED(src) || !active || charge < 5)
+		to_chat(usr, "<span class='warning'>The device is not able to play more DJ sounds at this time.</span>")
+		return
+	charge -= 5
+	playsound(src, S,300,1)
+
+/obj/machinery/disco/proc/check_GBP()
+	available |= "Engineering's Basic Beat"
+	available |= "Engineering's Domination Dance"
+	available |= "Engineering's Superiority Shimmy"
+	available |= "Engineering's Ultimate High-Energy Hustle"
+
+
+/obj/machinery/disco/proc/dance_setup()
+	stop = world.time + selection.song_length
+	var/turf/cen = get_turf(src)
+	FOR_DVIEW(var/turf/t, 3, get_turf(src),INVISIBILITY_LIGHTING)
+		if(t.x == cen.x && t.y > cen.y)
+			var/obj/item/device/flashlight/spotlight/L = new /obj/item/device/flashlight/spotlight(t)
+			L.light_color = "red"
+			L.light_power = 30-(get_dist(src,L)*8)
+			L.range = 1+get_dist(src, L)
+			spotlights+=L
+			continue
+		if(t.x == cen.x && t.y < cen.y)
+			var/obj/item/device/flashlight/spotlight/L = new /obj/item/device/flashlight/spotlight(t)
+			L.light_color = "purple"
+			L.light_power = 30-(get_dist(src,L)*8)
+			L.range = 1+get_dist(src, L)
+			spotlights+=L
+			continue
+		if(t.x > cen.x && t.y == cen.y)
+			var/obj/item/device/flashlight/spotlight/L = new /obj/item/device/flashlight/spotlight(t)
+			L.light_color = "#ffff00"
+			L.light_power = 30-(get_dist(src,L)*8)
+			L.range = 1+get_dist(src, L)
+			spotlights+=L
+			continue
+		if(t.x < cen.x && t.y == cen.y)
+			var/obj/item/device/flashlight/spotlight/L = new /obj/item/device/flashlight/spotlight(t)
+			L.light_color = "green"
+			L.light_power = 30-(get_dist(src,L)*8)
+			L.range = 1+get_dist(src, L)
+			spotlights+=L
+			continue
+		if((t.x+1 == cen.x && t.y+1 == cen.y) || (t.x+2==cen.x && t.y+2 == cen.y))
+			var/obj/item/device/flashlight/spotlight/L = new /obj/item/device/flashlight/spotlight(t)
+			L.light_color = "sw"
```

Comment:
There are only 4 colors that rotate around the disco ball. This designates the diagonals which are initially off and allows the color rotation to proceed.